File: General-DSA-Questions/Heap/9_KthLargestSumSubArray.cpp
// Write a function to find the kth largest sum of a subarray in an array.

// Given an array arr[] of size n and an integer k, find the kth largest sum of any contiguous subarray.
// Example:
// Input: arr[] = {20, -5, -1} , k = 3
// Contiguous subarrays and their sums are:
// 1. {20} -> 20
// 2. {20, -5} -> 15
// 3. {20, -5, -1} -> 14
// 4. {-5} -> -5
// 5. {-5, -1} -> -6
// 6. {-1} -> -1
// Sorting the sums gives us: {20, 15, 14, -1, -5, -6}
// The 3rd largest sum is 14.

// Approach: (Brute Force)
// 1. Generate all possible contiguous subarrays and calculate their sums.
// 2. Sort the sums in descending order.
// 3. Return the kth largest sum.
// 4. Time Complexity: O(N^2 log N) for generating subarrays and sorting.
// 5. Space Complexity: O(N) for storing the sums.

// Another Approach: (using min-heap):
// 1. Iterate all possible subarrays using two loops.
// 2. Compute their sum.
// 3. Maintain a min heap of size k:
//    a.  If heap size < k, push the sum.
//    b.  Else, compare and replace smallest if current sum is larger.
// 4. This way, after processing all subarrays, the top of the heap is the Kth largest.
// 5. Time Complexity: O(N^2 log K) for generating subarrays and maintaining the heap.
// 6. Space Complexity: O(K) for storing the k largest sums in the min-heap.

#include <iostream>
#include <queue>
using namespace std;

// Function to find the kth largest sum of a subarray
int kthLargestSumSubarray(int arr[], int n, int k) {
    priority_queue<int, vector<int>, greater<int>> minHeap; // Min-heap to store the k largest sums

    // Iterate through all possible subarrays
    for (int i = 0; i < n; i++) {
        int sum = 0; // Initialize sum for the current subarray
        for (int j = i; j < n; j++) {
            sum += arr[j]; // Calculate the sum of the current subarray

            // If heap size is less than k, push the sum
            if (minHeap.size() < k) {
                minHeap.push(sum);
            } else if (sum > minHeap.top()) { // If current sum is larger than the smallest in the heap
                minHeap.pop(); // Remove the smallest sum
                minHeap.push(sum); // Add the current sum
            }
        }
    }

    return minHeap.top(); // The top of the min-heap is the kth largest sum
}

int main() {
    int arr[] = {20, -5, -1}; // Example array
    int n = sizeof(arr) / sizeof(arr[0]);
    int k = 3; // Find the 3rd largest sum

    int result = kthLargestSumSubarray(arr, n, k);
    cout << "The " << k << "rd largest sum of a subarray is: " << result << endl;

    return 0;
}

File: JS-DSA-Questions/Heap/9_KthLargestSumSubArray.js
// Find the kth largest sum among all contiguous subarrays.

// Example: arr = [20, -5, -1], k = 3
// Subarray sums: 20, 15, 14, -5, -6, -1 → sorted desc: 20, 15, 14, -1, -5, -6 → 3rd largest = 14.

// Approach (min-heap of size k):
// 1. Enumerate all subarray sums with two loops.
// 2. Keep a min-heap of size k: if the heap has < k items push; else if the current sum is larger
//    than the heap's min, pop and push. The heap's root is the kth largest sum.
// Time: O(N^2 log K). Space: O(K).

// JavaScript note: no built-in priority queue, so we include a small numeric MinHeap.
class MinHeap {
  constructor() {
    this.a = [];
  }
  size() {
    return this.a.length;
  }
  peek() {
    return this.a[0];
  }
  push(v) {
    this.a.push(v);
    let i = this.a.length - 1;
    while (i > 0) {
      const p = Math.floor((i - 1) / 2);
      if (this.a[p] > this.a[i]) {
        [this.a[p], this.a[i]] = [this.a[i], this.a[p]];
        i = p;
      } else break;
    }
  }
  pop() {
    const top = this.a[0];
    const last = this.a.pop();
    if (this.a.length > 0) {
      this.a[0] = last;
      let i = 0;
      const n = this.a.length;
      while (true) {
        const l = 2 * i + 1;
        const r = 2 * i + 2;
        let s = i;
        if (l < n && this.a[l] < this.a[s]) s = l;
        if (r < n && this.a[r] < this.a[s]) s = r;
        if (s !== i) {
          [this.a[i], this.a[s]] = [this.a[s], this.a[i]];
          i = s;
        } else break;
      }
    }
    return top;
  }
}

function kthLargestSumSubarray(arr, k) {
  const minHeap = new MinHeap();
  const n = arr.length;

  for (let i = 0; i < n; i++) {
    let sum = 0;
    for (let j = i; j < n; j++) {
      sum += arr[j];
      if (minHeap.size() < k) {
        minHeap.push(sum);
      } else if (sum > minHeap.peek()) {
        minHeap.pop();
        minHeap.push(sum);
      }
    }
  }

  return minHeap.peek();
}

console.log("The 3rd largest sum of a subarray is:", kthLargestSumSubarray([20, -5, -1], 3)); // 14
